zkevm-circuits/src/evm_circuit/execution/addmod.rs

```diff
@@ -0,0 +1,296 @@
+            ..StepStateTransition::default()
+        };
+        let same_context = SameContextGadget::construct(cb, opcode, step_state_transition);
+
+        Self {
+            a,
+            b,
+            r,
+            n,
+            k,
+            d,
+            a_reduced,
+            muladd_d_n_r,
+            muladd_k_n_areduced,
+            same_context,
+            cmp_r_n,
+            cmp_areduced_n,
+            n_is_zero,
+            sum_areduced_b,
+            sum_areduced_b_overflow,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _: &Transaction,
+        _: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        // get stack values
+        let [mut r, n, b, a] = [3, 2, 1, 0]
+            .map(|idx| step.rw_indices[idx])
+            .map(|idx| block.rws[idx].stack_value());
+
+        // assing a,b & n stack values
+        self.a.assign(region, offset, Some(a.to_le_bytes()))?;
+        self.b.assign(region, offset, Some(b.to_le_bytes()))?;
+        self.n.assign(region, offset, Some(n.to_le_bytes()))?;
+
+        // compute a_reduced,k,d,a_reduced_plus_b,a_reduced_plus_b_overflow,r values
+        let a_reduced;
+        let k;
+        let d;
+        let a_reduced_plus_b;
+        let a_reduced_plus_b_overflow;
+
+        if n.is_zero() {
+            k = U256::zero();
+            a_reduced = a;
+            d = U256::zero();
+            r = a.overflowing_add(b).0;
+
+            a_reduced_plus_b = a_reduced.overflowing_add(b).0;
+            a_reduced_plus_b_overflow = U256::zero();
+        } else {
+            let (a_div_n, a_mod_n) = a.div_mod(n);
+            k = a_div_n;
+            a_reduced = a_mod_n;
+            d = (a_reduced + b) / n;
+
+            let (sum, overflow) = a_reduced.overflowing_add(b);
+            a_reduced_plus_b = sum;
+            a_reduced_plus_b_overflow = if overflow { U256::one() } else { U256::zero() };
+        };
+
+        // rest of values and gadgets
+
+        self.r.assign(region, offset, Some(r.to_le_bytes()))?;
+        self.k.assign(region, offset, Some(k.to_le_bytes()))?;
+        self.d.assign(region, offset, Some(d.to_le_bytes()))?;
+        self.a_reduced
+            .assign(region, offset, Some(a_reduced.to_le_bytes()))?;
+
+        self.muladd_k_n_areduced
+            .assign(region, offset, [k, n, a_reduced, a])?;
+
+        self.sum_areduced_b
+            .assign(region, offset, [a_reduced, b], a_reduced_plus_b)?;
+
+        self.sum_areduced_b_overflow.assign(
+            region,
+            offset,
+            Some(a_reduced_plus_b_overflow.to_le_bytes()),
+        )?;
+        self.muladd_d_n_r.assign(
+            region,
+            offset,
+            [d, n, a_reduced_plus_b, a_reduced_plus_b_overflow],
```

Comment:
```suggestion
            [d, n, a_reduced_plus_b_overflow. a_reduced_plus_b],
```
The overflow bit should come first

Reply:
oh... We need to figure out why! :female_detective:

Reply:
I think I found the reason why this works when passing swapped `d` and `e` in the assign function of `MulAddWords512Gadget`.
On one hand, `MulAddWords512Gadget` only assigns to its internal witness which are `carry_0`, `carry_1` and `carry_2`.  And those values are calculated from `a, b, c, d, e`.  If I remember correctly, for ADDMOD we only needed 257 bit arithmetic, but we're using `MulAddWords512Gadget` for convenience.  This means that the possible values we will encounter for the carry witnesses is more limited (I think they are always 0 or 1).  And it seems that swapping `d` and `e` (using values from ADDMOD), the calculated carry witnesses stay the same!  Or at least that's the case with the tests.

I'm not sure if there's a case where swapping `d` and `e` in the assign function for `MulAddWords512Gadget` could change the carry witnesses.

Reply:
> I'm not sure if there's a case where swapping `d` and `e` in the assign function for `MulAddWords512Gadget` could change the carry witnesses.

I've tried to find some combination of inputs that calculates in different carry witnesses when swapping `d` and `e`, but I didn't find any.  It's a bit tricky because at this point the inputs are derived from other calculations.  So even after trying I'm not convinced that such inputs don't exist.

Anyway, I think we can merge this PR even if this is not yet resolved!